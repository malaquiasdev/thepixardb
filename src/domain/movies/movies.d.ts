type Image = {
  width: number;
  url: string;
  height: number;
};

export type MovieModel = {
  id: string;
  language: string;
  title: string;
  description: string;
  genres: string;
  adult: boolean;
  originalLanguage: string;
  runtime: number;
  budget: number;
  releaseYear: string;
  revenue?: number;
  updateAt: string;
  createAt: string;
  themoviedbId: number;
  imdbid: string;
  posters: Image[];
};

export type MovieDTO = {
  id: string;
  title: string;
  description: string;
  genres: string[];
  adult: boolean;
  originalLanguage: string;
  runtime: number;
  budget: number;
  releaseYear: string;
  revenue: number;
  posters: Image[];
};

export type MovieListDTO = {
  id: string;
  title: string;
  releaseYear: string;
  banner?: Image;
  poster?: Image;
};
