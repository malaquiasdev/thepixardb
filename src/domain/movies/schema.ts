export const moviesSchema = {
  tags: ['movies'],
  description: 'Get the primary information about a movie.',
  querystring: {
    type: 'object',
    additionalProperties: false,
    properties: {
      language: {
        type: 'string',
        description: 'Pass a ISO 639-1 value to display translated data for the fields that support it.',
        default: 'en',
        enum: ['en', 'pt'],
      },
      genre: {
        type: 'string',
      },
    },
  },
  response: {
    200: {
      type: 'array',
      items: {
        type: 'object',
        properties: {
          id: { type: 'string' },
          title: { type: 'string' },
          description: { type: 'string' },
          releaseYear: { type: 'string' },
          runtime: { type: 'number' },
          posterUrl: { type: 'string' },
          bannerUrl: { type: 'string' },
        },
      },
    },
  },
};

export const movieSchema = {
  tags: ['movies'],
  description: '',
  params: {
    type: 'object',
    properties: {
      id: {
        type: 'string',
        description: 'movie id',
      },
    },
  },
  querystring: {
    type: 'object',
    additionalProperties: false,
    properties: {
      language: {
        type: 'string',
        description: 'Pass a ISO 639-1 value to display translated data for the fields that support it.',
        default: 'en',
        enum: ['en', 'pt'],
      },
    },
  },
  response: {
    200: {
      type: 'object',
      properties: {
        id: { type: 'string' },
        title: { type: 'string' },
        description: { type: 'string' },
        releaseYear: { type: 'string' },
        runtime: { type: 'number' },
        poster: {
          type: 'object',
          properties: {
            width: { type: 'integer' },
            height: { type: 'integer' },
            url: { type: 'string' },
          },
        },
      },
    },
  },
};
