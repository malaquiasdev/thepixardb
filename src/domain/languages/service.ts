export function listAllowedLanguages() {
  return [
    {
      iso6391: 'JA',
      description: 'Japanese',
    },
    {
      iso6391: 'EN_US',
      description: 'English (United States)',
    },
    {
      iso6391: 'PT_BR',
      description: 'Portuguese (Brazil)',
    },
  ];
}
