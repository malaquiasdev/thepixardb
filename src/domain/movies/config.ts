export const config = {
  TABLE_MOVIES: process.env.TABLE_MOVIES,
  CACHE_MAX_AGE: 604800,
};
