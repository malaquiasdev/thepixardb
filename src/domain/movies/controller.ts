import { FastifyReply } from 'fastify';
import { findMovies, findMovie } from './service';
import { config } from './config';

export async function moviesController(req: any, reply: FastifyReply): Promise<FastifyReply> {
  const data = await findMovies(config.TABLE_MOVIES, req.query.language);
  return reply
    .headers({
      'Cache-Control': `max-age=${config.CACHE_MAX_AGE}`,
      'Access-Control-Allow-Origin': '*',
    })
    .send(data);
}

export async function movieController(req: any, reply: FastifyReply): Promise<FastifyReply> {
  const data = await findMovie(config.TABLE_MOVIES, req.params.id, req.query.language);
  return reply
    .headers({
      'Cache-Control': `max-age=${config.CACHE_MAX_AGE}`,
      'Access-Control-Allow-Origin': '*',
    })
    .send(data);
}
