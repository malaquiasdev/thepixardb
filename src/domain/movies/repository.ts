import { executePartiQL } from '../../aws/dynamo';
import { MovieModel } from './movies';

export async function findMoviesBy(tableName: string, language?: string): Promise<MovieModel[]> {
  const query = `SELECT * FROM ${tableName} WHERE "sk" = 'LANGUAGE#${language}' `;
  return await executePartiQL(query);
}

export async function findMovieBy(tableName: string, id: string, language?: string): Promise<MovieModel> {
  const query = `SELECT * FROM ${tableName} WHERE "pk" = '${id}' AND "sk" = 'LANGUAGE#${language}'`;
  const result = await executePartiQL(query);
  return result[0];
}
