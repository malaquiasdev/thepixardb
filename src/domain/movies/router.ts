import fp from 'fastify-plugin';
import { moviesController, movieController } from './controller';
import { moviesSchema, movieSchema } from './schema';

export default fp(async (server, _, next) => {
  server.route({
    url: '/movies',
    logLevel: 'warn',
    method: ['GET', 'HEAD', 'OPTIONS'],
    handler: moviesController,
    moviesSchema,
  });
  server.route({
    url: '/movies/:id',
    logLevel: 'warn',
    method: ['GET', 'HEAD', 'OPTIONS'],
    handler: movieController,
    movieSchema,
  });
  next();
});
