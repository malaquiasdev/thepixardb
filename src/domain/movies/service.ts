import { findMoviesBy, findMovieBy } from './repository';
import { MovieDTO, MovieListDTO } from './movies';

function _getImageUrl(images, ratio: string, language: string): string {
  const image = images.find(image => image.key === language && image.path.includes(`/${ratio}/${language}`));

  if (image) {
    return image.path;
  }

  return '';
}

export async function findMovies(tableName: string, language = 'en'): Promise<MovieListDTO[]> {
  const movies = await findMoviesBy(tableName, language);
  return movies.map(m => {
    return {
      id: m.pk,
      title: m.title,
      description: m.descriptionShort,
      releaseDate: m.releaseDate,
      runtime: m.runtime,
      posterUrl: _getImageUrl(m.images, '2_3', language),
      bannerUrl: _getImageUrl(m.images, '16_9', language),
    };
  });
}

export async function findMovie(tableName: string, id: string, language = 'en'): Promise<MovieDTO> {
  const movie = await findMovieBy(tableName, id, language);
  return {
    id: movie.pk,
    title: movie.title,
    description: movie.description,
    descriptionShort: movie.descriptionShort,
    genres: movie.genres,
    runtime: movie.runtime,
    budget: movie.budget,
    releaseDate: movie.releaseDate,
    revenue: movie.revenue,
    rating: movie.rating,
    images: movie.images.filter(i => i.primary && i.key === language),
  };
}
