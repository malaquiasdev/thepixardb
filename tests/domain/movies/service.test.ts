import * as repository from '../../../src/domain/movies/repository';
import * as service from '../../../src/domain/movies/service';
import movies from './__mocks__/movies.en.mock.json';

jest.mock('../../../src/domain/movies/repository');

describe('findMovies', () => {
  beforeEach(() => {
    (repository.findMoviesBy as jest.Mock).mockReset();
    (repository.findMovieBy as jest.Mock).mockReset();
    (repository.findMoviesBy as jest.Mock).mockResolvedValue(movies);
    (repository.findMovieBy as jest.Mock).mockResolvedValue(movies[0]);
  });
  it('should return movies', async () => {
    const result = await service.findMovies('tableName', 'en');
    expect(result).toBeInstanceOf(Array);
    expect(result).toMatchSnapshot();
  });
});

describe('findMovie', () => {
  it('should return movie', async () => {
    const result = await service.findMovie('tableName', 'MV508439', 'en');
    expect(result).toBeInstanceOf(Object);
    expect(result).toMatchSnapshot();
  });
});
