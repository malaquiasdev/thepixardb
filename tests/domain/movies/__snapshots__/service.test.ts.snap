// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`findMovie should return movie 1`] = `
Object {
  "budget": 200000000,
  "description": "In a suburban fantasy world, two teenage elf brothers embark on an extraordinary quest to discover if there is still a little magic left out there.",
  "descriptionShort": "Two elven brothers embark on a quest to bring their father back for one day.",
  "genres": Array [
    "Animation",
    "Family",
    "Comedy",
    "Fantasy",
    "Adventure",
  ],
  "id": "MV508439",
  "images": Array [
    Object {
      "key": "en",
      "path": "https://raw.githubusercontent.com/malaquiasdev/thepixardb-wall-e/main/data/movies/MV508439/images/2_3/en_w2000_h3000.jpeg",
      "primary": true,
    },
    Object {
      "key": "en",
      "path": "https://raw.githubusercontent.com/malaquiasdev/thepixardb-wall-e/main/data/movies/MV508439/images/16_9/en_w1440_h810.jpeg",
      "primary": true,
    },
  ],
  "rating": Object {
    "key": "en",
    "organ": "Motion Picture Association of America",
    "value": "PG",
  },
  "releaseDate": "2020-02-29",
  "revenue": 141592359,
  "runtime": 103,
  "title": "Onward",
}
`;

exports[`findMovies should return movies 1`] = `
Array [
  Object {
    "bannerUrl": "https://raw.githubusercontent.com/malaquiasdev/thepixardb-wall-e/main/data/movies/MV508439/images/16_9/en_w1440_h810.jpeg",
    "description": "Two elven brothers embark on a quest to bring their father back for one day.",
    "id": "MV508439",
    "posterUrl": "https://raw.githubusercontent.com/malaquiasdev/thepixardb-wall-e/main/data/movies/MV508439/images/2_3/en_w2000_h3000.jpeg",
    "releaseDate": "2020-02-29",
    "runtime": 103,
    "title": "Onward",
  },
  Object {
    "bannerUrl": "https://raw.githubusercontent.com/malaquiasdev/thepixardb-wall-e/main/data/movies/MV260514/images/16_9/en_w1920_h1080.jpeg",
    "description": "Lightning McQueen sets out to prove to a new generation of racers that he's still the best race car in the world.",
    "id": "MV260514",
    "posterUrl": "https://raw.githubusercontent.com/malaquiasdev/thepixardb-wall-e/main/data/movies/MV260514/images/2_3/en_w1200_h1800.jpeg",
    "releaseDate": "2017-06-15",
    "runtime": 90,
    "title": "Cars 3",
  },
  Object {
    "bannerUrl": "https://raw.githubusercontent.com/malaquiasdev/thepixardb-wall-e/main/data/movies/MV863/images/16_9/en_w1920_h1080.jpeg",
    "description": "When Woody is stolen by a toy collector, Buzz and his friends set out on a rescue mission to save Woody before he becomes a museum toy property with his roundup gang Jessie, Prospector, and Bullseye.",
    "id": "MV863",
    "posterUrl": "https://raw.githubusercontent.com/malaquiasdev/thepixardb-wall-e/main/data/movies/MV863/images/2_3/en_w800_h1171.jpeg",
    "releaseDate": "1999-10-30",
    "runtime": 92,
    "title": "Toy Story 2",
  },
]
`;
